tree/dataframe/src/RCsvDS.cxx

```diff
@@ -123,18 +123,18 @@ void RCsvDS::FillRecord(const std::string &line, Record_t &record)
 
       switch (colType) {
       case 'd': {
-         record.emplace_back(new double(std::stod(col)));
+         record.emplace_back(new double((col != "NaN") ? std::stod(col) : std::numeric_limits<double>::quiet_NaN()));
          break;
       }
       case 'l': {
-         record.emplace_back(new Long64_t(std::stoll(col)));
+         record.emplace_back(
+            new Long64_t((col != "NaN") ? std::stoll(col) : std::numeric_limits<Long64_t>::quiet_NaN()));
          break;
       }
       case 'b': {
          auto b = new bool();
          record.emplace_back(b);
-         std::istringstream is(col);
-         is >> std::boolalpha >> *b;
+         std::istringstream((col != "NaN") ? col : "false") >> std::boolalpha >> *b;
```

Comment:
So here we lose the information that the value was missing by transforming it to false. How does pandas treat this case?

Reply:
I agree, but it seems to be no workaround it. Using a `quiet_NaN` would again return me `false` - https://en.cppreference.com/w/cpp/types/numeric_limits/quiet_NaN. `NaN` is a thing only for `float`, `double`, `long double` (see link).

Pandas puts `NaN` everywhere it sees an empty cell. And it produces vectors of objects, i.e. a column might have integer and a string inside it.

Numpy reads accepts only numerical values and represent them as floats. On the place of all other input cells (strings, chars, or empty cell) is `nan`.

Provided that we have support for integers and booleans, which do not have actual `NaN` values, we might warn the user in my opinion.

Reply:
So internally we can only implement NaNs for floating point numbers. That's not ideal since we can't be consistent across types - for some types we lose the information that there was actually no value. @eguiraud do you think that, for the purposes of RDataFrame, it's enough if we can e.g. filter rows where the value of a floating point column is a NaN, but we can't tell if the value was not available for e.g. ints or bools? That would at least cover what https://root-forum.cern.ch/t/nan-values-in-one-column-break-histogram-of-another-column/37091/3 requested.

Reply:
it might be good enough. more importantly, I don't know what else we can do :sweat_smile: 

related: what do we do for empty values of other types? throw an exception, fill in a default..?j

Reply:
My current commit is using corresponding `quiet_NaN` for other types as well, which is "bad" for integers/booleans.

An option that I think now is: If a current column was initially determined as integer or boolean, and it contains an empty cell (for which no specific `fillValue` is specified - so we need `NaN`-s), I can cast the whole column to floats and properly put "floating `nan`" where needed.

Reply:
@etejedor @eguiraud A more efficient suggestion that I have just realized now is:

1. Do not change anything of the current csv reader functionality. Only make an explicit check for empty cells, which will tell the user to use `MakeNumericalCsvDataframe`
2. Introduce a new method `MakeNumericalCsvDataframe`, which will read every column as double. Only `MakeNumericalCsvDataframe` will allow parsing csv files with empty cells. By default empty cells will be `quite_NaN<double>`(hence "real" `nan`), and since every entry is a double, a user will be able to define a fill value. In fact, doubles should cover the usecases for integers (except very large ones) and boolean values.
3. In this way we will "mock" the behaviour of csv reader in Numpy.

Reply:
It seems too much to have to maintain a separate data-source implementation just for this. For now it is enough to raise a user-friendly error in case a non-floating-point value is missing, that could also suggest to mark that column as floating point to allow NaN values.

Reply:
I agree with what @eguiraud said, but regarding:

>  that could also suggest to mark that column as floating point to allow NaN values.

we currently don't offer any mechanism for the user to specify the type of a column, do we? We'd need to add that as well.

Reply:
Okay, I agree as well. Regarding the type specifications, I am thinking of 2 ways.

First, similar to the `TTree::ReadFile`, user needs to add to their input file's header line "/F" to specify column is float, and so on for other types. So if a column was called "Pt", user need to modify it to "Pt/F". And we can "hide" the type hint ("/F") when later asked to show the particular column name.

Alternative, which would not require the user to modify the input file is to pass additional argument to the csv dataframe, which is most likely a vector of pairs: left element of the pair is column index, and right element is the name of the desired type. For example, one can call `MakeCsvDataframe(... , customColTypes = {{3, "double"}, {4, "bool"}})`, to apply these custom column types. This could be done as vector of strings where each string is of the type `"colIdx:colType"`. The idea is that the csv dataframe supports parsing files without column names, and this solution easily covers that.